src/alphafold_impact/pipelines/data_collection_oa/nodes.py

```diff
@@ -0,0 +1,117 @@
+import requests
+from tqdm import tqdm
+import math
+from typing import List
+import json
+from kedro_datasets.json import JSONDataset
+
+
+def retrieve_oa_works_for_single_concept_and_year(
+    concept_id: str, year: int, per_page: int
+) -> List[dict]:
+    """
+    Retrieves OpenAlex works for a single concept ID and publication year.
+
+    Args:
+        concept_id (str): The OpenAlex concept ID.
+        year (int): The publication year.
+        per_page (int): The number of results to retrieve per page.
+
+    Returns:
+        List[dict]: A list of works for the given concept ID and year.
+    """
+    base_url = "https://api.openalex.org/works"
+    works = []
+    next_cursor = "*"
+    first_request = True
+
+    while True:
+        params = {
+            "filter": f"concepts.id:{concept_id},publication_year:{year}",
+            "per_page": per_page,
+            "cursor": next_cursor,
+        }
+        response = requests.get(base_url, params=params)
+        if response.status_code == 200:
+            data = response.json()
+            current_works = data.get("results", [])
+            if first_request:
+                # Calculate the total number of pages from the first response
+                total_count = data.get("meta", {}).get("count", 0)
+                total_pages = math.ceil(total_count / per_page)
+                page_progress = tqdm(
+                    total=total_pages, desc="Pages", leave=False
+                )  # Initialize progress bar with total
+                first_request = False
+
+            if not current_works:  # Break if there are no more works
+                break
+            works.extend(current_works)
+
+            # Update the page progress bar
+            page_progress.update(1)
+
+            # Retrieve the next cursor from the response
+            next_cursor = data.get("meta", {}).get("next_cursor")
+            if not next_cursor:  # Break if there's no next cursor
+                break
+        else:
+            print(
+                f"Error fetching data for concept {concept_id} and year {year}: {response.status_code}"
+            )
+            break
```

Comment:
Is this behaviour not a bit risky? If OpenAlex refuses 200 handshake for any temporary reason, you exit the loop. It may be good to instist a bit by keeping a session open and retrying a handful of times. I haven't had issues with OA on this, but for example it happens all the time with GtR. You can check #11 for the approach I take, namely:

- import adapters `from requests.adapters import HTTPAdapter, Retry`  
- sustain session: 
`url = f"{base_url}{endpoint}?p={page}&s={page_size}"`
`session = requests.Session()`
`retries = Retry(total=max_retries, backoff_factor=backoff_factor)`
`session.mount("https://", HTTPAdapter(max_retries=retries))`
`response = session.get(url, headers=headers, timeout=30)`

Reply:
Thanks, I've added this functionality now.